src/routers/bookmark_router.py

```diff
@@ -0,0 +1,123 @@
+from fastapi import APIRouter, Depends, HTTPException
+from typing import List, Union
+from pydantic import BaseModel, validator
+from sqlmodel import Session, select, Field
+
+from authentication import KeycloakUser, get_user_or_raise
+from database.session import get_session
+from database.model.bookmark.bookmark import Bookmark
+from http import HTTPStatus
+from database.model.concept.concept import AIoDConcept
+from database.model.helper_functions import non_abstract_subclasses
+from datetime import datetime
+from routers.helper_functions import get_asset_type_by_abbreviation
+
+
+class BookmarkRead(BaseModel):
+    resource_identifier: str = Field(description="The identifier of the resource being bookmarked.")
+    created_at: datetime = Field(description="The time when the bookmark was created in ISO 8601 format.")
+
+    class Config:
+        json_encoders = {datetime: lambda dt: dt.isoformat()}
+
+
+def create(url_prefix: str = "") -> APIRouter:
+    router = APIRouter()
+
+    for path in [
+        f"{url_prefix}/v2/bookmarks",
+        f"{url_prefix}/bookmarks",
+    ]:
+
+        @router.get(
+            path,
+            tags=["User"],
+            description="Return all assets you have bookmarked.",
+            response_model=List[BookmarkRead],
+        )
+        def list_bookmarks(
+            user: KeycloakUser = Depends(get_user_or_raise), session: Session = Depends(get_session)
+        ) -> List[BookmarkRead]:
+            return session.exec(
+                select(Bookmark).where(Bookmark.user_identifier == user._subject_identifier)
+            ).all()
+
+        @router.post(
+            path,
+            tags=["User"],
+            response_model=BookmarkRead,
+            description="Add the asset to the logged-in user's bookmarks.",
+            status_code=HTTPStatus.OK,
+        )
+        def create_bookmark(
+            resource_identifier: str,
+            user: KeycloakUser = Depends(get_user_or_raise),
+            session: Session = Depends(get_session),
+        ) -> BookmarkRead:
+            # # Check if the resource exists
+            if not resource_identifier_exists_in_database(resource_identifier, session):
+                raise HTTPException(
+                    status_code=HTTPStatus.NOT_FOUND,
+                    detail=f"Resource {resource_identifier} does not exist.",
+                )
+
+            # Prevent duplicate bookmarks
+            db_bookmark = session.exec(
+                select(Bookmark).where(
+                    Bookmark.user_identifier == user._subject_identifier,
+                    Bookmark.resource_identifier == resource_identifier,
+                )
+            ).first()
```

Comment:
```suggestion
            db_bookmark = session.get(Bookmark, (user._subject_identifier, resource_identifier))
```
For fetching something by their primary key(s), using the `session.get` method is much clearer! Something that unfortunately this project doesn't use much (yet).